tensorboard/webapp/widgets/linked_time_fob/linked_time_fob_test.ts

```diff
@@ -0,0 +1,319 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+import {Component, NO_ERRORS_SCHEMA, ViewChild} from '@angular/core';
+import {ComponentFixture, TestBed} from '@angular/core/testing';
+import {LinkedTime} from '../../metrics/types';
+import {ScaleLinear, ScaleTime} from '../../third_party/d3';
+import {LinkedTimeFobComponent} from './linked_time_fob_component';
+import {
+  AxisDirection,
+  Fobs,
+  LinkedTimeFobControllerComponent,
+} from './linked_time_fob_controller_component';
+
+type TemporalScale = ScaleLinear<number, number> | ScaleTime<number, number>;
+
+@Component({
+  selector: 'testable-comp',
+  template: `
+    <linked-time-fob-controller
+      #FobController
+      [axisDirection]="axisDirection"
+      [linkedTime]="linkedTime"
+      [steps]="steps"
+      [temporalScale]="temporalScale"
+      (onSelectTimeChanged)="onSelectedTimeChangedMock($event)"
+    ></linked-time-fob-controller>
+  `,
+})
+class TestableComponent {
+  @ViewChild('FobController')
+  fobController!: LinkedTimeFobControllerComponent;
+
+  steps!: number[];
+  axisDirection!: AxisDirection;
+  linkedTime!: LinkedTime;
+  temporalScale!: TemporalScale;
+
+  onSelectedTimeChangedMock(newLinkedTime: LinkedTime) {
+    this.linkedTime = newLinkedTime;
+  }
+}
+
+describe('linked_time_fob_controller', () => {
+  let newLinkedTimeActionSpy: jasmine.Spy;
+  let temporalScaleSpy: jasmine.Spy;
+  beforeEach(async () => {
+    await TestBed.configureTestingModule({
+      declarations: [
+        TestableComponent,
+        LinkedTimeFobControllerComponent,
+        LinkedTimeFobComponent,
+      ],
+      schemas: [NO_ERRORS_SCHEMA],
+    }).compileComponents();
+  });
+
+  function createComponent(input: {
+    steps?: number[];
+    axisDirection?: AxisDirection;
+    linkedTime?: LinkedTime;
+  }): ComponentFixture<TestableComponent> {
+    const fixture = TestBed.createComponent(TestableComponent);
+    fixture.componentInstance.steps = input.steps || [1, 2, 3, 4];
+
+    fixture.componentInstance.axisDirection =
+      input.axisDirection || AxisDirection.VERTICAL;
+
+    fixture.componentInstance.linkedTime = input.linkedTime || {
+      start: {step: 1},
+      end: null,
+    };
+
+    temporalScaleSpy = jasmine.createSpy();
+    fixture.componentInstance.temporalScale =
+      temporalScaleSpy as unknown as ScaleLinear<number, number>;
+
+    temporalScaleSpy.and.callFake((step: number) => {
+      return step;
+    });
+
+    newLinkedTimeActionSpy = jasmine.createSpy();
+    fixture.componentInstance.onSelectedTimeChangedMock =
+      newLinkedTimeActionSpy;
+    newLinkedTimeActionSpy.and.stub();
+
+    return fixture;
+  }
+
+  it('moves the start fob down to mouse when mouse is dragging down and is below fob', () => {
```

Comment:
Does this mean the mouse is not dragging the fob "in a correct way"? What looks to me is, the mouse is below the fob a bit but still touches the fob and the mouse is moving up. This should also move the fob. If not, could you explain more on what this test is about. Under what scenario it happens?